state/storage.go

```diff
@@ -306,6 +306,77 @@ func (s *Storage) MetadataForAllRooms(txn *sqlx.Tx, tempTableName string, result
 	return nil
 }
 
+// ResetMetadataState updates the given metadata in-place to reflect the current state
+// of the room.
+// TODO: could be brittle. Might be better to just create a new RoomMetadata from scratch
+// and copy over the fields we don't want to reload.
+func (s *Storage) ResetMetadataState(metadata *internal.RoomMetadata) error {
+	var events []Event
+	err := s.DB.Select(&events, `
+	WITH snapshot(events, membership_events) AS (
+        SELECT events, membership_events
+        FROM syncv3_snapshots
+            JOIN syncv3_rooms ON snapshot_id = current_snapshot_id
+        WHERE syncv3_rooms.room_id = $1
+    )
+	SELECT event_id, event_type, state_key, event, membership
+	FROM syncv3_events JOIN snapshot ON (
+		event_nid = ANY (ARRAY_CAT(events, membership_events))
+	)
+	WHERE (event_type IN ('m.room.name', 'm.room.avatar', 'm.room.canonical_alias') AND state_key = '')
+	   OR (event_type = 'm.room.member' AND membership IN ('join', '_join', 'invite', '_invite'))
+	ORDER BY event_nid ASC
+	;`, metadata.RoomID)
+	if err != nil {
+		return fmt.Errorf("ResetMetadataState[%s]: %w", metadata.RoomID, err)
+	}
+
+	heroMemberships := circularSlice[*Event]{max: 6}
```

Comment:
Note: when populating the heroes I only care about the 6 most recent heroes. However to get accurate join and invite counts we effectively need to pull all the members out.

Reply:
If we're only worrying about redactions then there aren't gonna be any count changes. But as #294 says there's the more general problem of "our state is completely out of date".